employee-rostering-frontend/src/ui/components/DataTable.tsx

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React from 'react';
+import {
+  Table,
+  TableHeader,
+  TableBody,
+  headerCol,
+  IRow
+} from '@patternfly/react-table';
+import {Button, ButtonVariant} from '@patternfly/react-core';
+import { SaveIcon, CloseIcon, EditIcon, TrashIcon } from '@patternfly/react-icons';
+import { EditableComponent } from './EditableComponent';
+
+export interface DataTableProps<T,C> {
+  title: string;
+  columnTitles: string[];
+  tableData: T[];
+}
+
+interface DataTableState<T> {
+  newRow: IRow | null;
+  currentFilter: (rowData: T) => boolean;
+}
+
+export abstract class DataTable<T,C, P extends DataTableProps<T,C>> extends React.Component<P, DataTableState<T>> {
+  constructor(props: P) {
+    super(props);
+    this.createNewRow = this.createNewRow.bind(this);
+    this.cancelAddingRow = this.cancelAddingRow.bind(this);
+    this.convertDataToTableRow = this.convertDataToTableRow.bind(this);
+    this.state = {newRow: null, currentFilter: (t) => true};
+  }
+
+  abstract displayDataRow(data: T): JSX.Element[];
+  abstract editDataRow(dataStore: C, data: T): JSX.Element[];
+  abstract createNewDataRow(dataStore: C): JSX.Element[];
+  abstract isValid(editedValue: C): boolean;
+  abstract extractDataFromRow(oldValue: T|{}, editedValue: C): T;
+
+  abstract updateData(data: T): void;
+  abstract addData(data: T): void;
+  abstract removeData(data: T): void;
+
+  createNewRow() {
+    if (this.state.newRow === null) {
+      const dataStore: any = {};
+      const newRowComponents: JSX.Element[] = this.createNewDataRow(dataStore);
+      const newRow: IRow = { cells: newRowComponents.map(c => {return {title: c}}).concat([{
+        title: this.getAddButtons(dataStore)}])};
+      this.setState({...this.state, newRow: newRow});
```

Comment:
I believe that component state should only contain simple types (strings, numbers, booleans, data structures made of these simple types) and not JSX elements, React nodes etc. Those should be created only when `render()` is called by React based on state or props changes.

Non-blocking suggestion. We can improve this later.

Reply:
My concern is losing the data on the new Row if the state changes. Maybe I can change the property to `newRowData: C|null` (where `C` is composed of simple types).